sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/BulkExecutor.java

```diff
@@ -0,0 +1,284 @@
+                return this.executePartitionKeyRangeServerBatchRequest(serverRequest, groupSink);
+            });
+    }
+
+    private Flux<CosmosBulkOperationResponse<TContext>> executePartitionKeyRangeServerBatchRequest(
+        PartitionKeyRangeServerBatchRequest serverRequest,
+        FluxSink<CosmosItemOperation> groupSink) {
+
+        return this.executeBatchRequest(serverRequest)
+            .flatMapMany(response -> {
+
+                return Flux.fromIterable(response.getResults()).flatMap((TransactionalBatchOperationResult result) -> {
+                    return handleTransactionalBatchOperationResult(response, result, groupSink);
+                });
+            })
+            .onErrorResume((Throwable throwable) -> {
+
+                return Flux.fromIterable(serverRequest.getOperations()).flatMap((CosmosItemOperation itemOperation) -> {
+                    return handleTransactionalBatchExecutionException(itemOperation, throwable, groupSink);
+                });
+            });
+    }
+
+    // Helper functions
+    private Mono<CosmosBulkOperationResponse<TContext>> handleTransactionalBatchOperationResult(
+        TransactionalBatchResponse response,
+        TransactionalBatchOperationResult operationResult,
+        FluxSink<CosmosItemOperation> groupSink) {
+
+        CosmosBulkItemResponse cosmosBulkItemResponse = BridgeInternal.createCosmosBulkItemResponse(operationResult, response);
+        CosmosItemOperation itemOperation = operationResult.getOperation();
+
+        if (!operationResult.isSuccessStatusCode()) {
+
+            if(itemOperation instanceof ItemBulkOperation<?>) {
+
+                return ((ItemBulkOperation<?>) itemOperation).getRetryPolicy().shouldRetry(operationResult).flatMap(
+                    result -> {
+                        if (result.shouldRetry) {
+                            groupSink.next(itemOperation);
+                            return Mono.empty();
+                        } else {
+                            return Mono.just(BridgeInternal.createCosmosBulkOperationResponse(
+                                itemOperation, cosmosBulkItemResponse, this.batchContext));
+                        }
+                    });
+
+            } else {
+                throw new UnsupportedOperationException("Unknown CosmosItemOperation.");
+            }
+        }
+
+        return Mono.just(BridgeInternal.createCosmosBulkOperationResponse(
+            itemOperation,
+            cosmosBulkItemResponse,
+            this.batchContext));
+    }
+
+    private Mono<CosmosBulkOperationResponse<TContext>> handleTransactionalBatchExecutionException(
+        CosmosItemOperation itemOperation,
+        Throwable throwable,
+        FluxSink<CosmosItemOperation> groupSink) {
+
+        if(throwable instanceof CosmosException && itemOperation instanceof ItemBulkOperation<?>) {
+            CosmosException cosmosException = (CosmosException) throwable;
```

Comment:
nit: if (...) (space between)

Reply:
I can see both "if( " ~ without space and "if (" ~ with spaces in code. Not sure what we follow.

Reply:
Will change it. The "if (" seems to be widely accepted elsewhere. Thanks.